### ISSUE TITLE
Analysis Ticket for #1749 - Handling of full time federal life time maximum restrictions.

### ISSUE THREAD

**Issue by JasonCTang** [bcgov/SIMS#1768]
**This is an ANALYSIS ticket only. **
The purpose of this ticket is to perform analysis for #1749.
Expectation is to solidfy the approach, have a better understanding of the requirements and see clarification from the business where needed.

**Primary Goals** 
 - Determine the best database approach to record restriction information associated with changes to the disbursement.
 - Determine logical steps for applying the award deduction based on the restriction and raise questions to business area.
 - Determine how we know which restriction to apply and the order that we through all restrictions and raise questions to business area.

**Description copied from #1749**
Federal Government is in charge of tracking a student against their fulltime federal lifetime maximums and they will notify us when to apply a restriction or not. In other words, there is no need to perform any calculation to reduce an award at time of ECERT generation. If funding is restricted, the award amount will just become zero.

Additional context

When a student has Full Time Federal Lifetime Maximums has been reached restriction, it stops funding to both federal and provincial funding. Essentially, all disbursements are stopped. (This action is already set in the restrictions table).
These restrictions only apply to Fulltime Application Disbursements.
Depending on the program, credential and PD status, it is possible to have a lifetime max restriction in one scenario but still be allowed funding in a different scenario. Rules are as follows:
If the application is for an Aviation program, then the AV restriction supercedes all other restrictions.
If the student has PD status and the application is NOT for an Aviation program, then the restriction 5, B7, supercedes the doctorate and non-doctorate restrictions.
If the student does NOT have PD status and the application is not for an Aviation program, then applicable restriction depends on whether the application is for a doctorate or non-doctorate credentials.
According to the Restrictions Analysis file, the restriction code B5 is no longer needed when we switch to SIMS.
Logic for when and how to apply federal restriction codes
At time of ECERT generation:
if (Application = Aviation Program & Restriction AV= True) {stop disbursements & record reason}
else if (PD Status = True )
{  
if (Restriction 5 = True or Restriction B7 = True) {stop disbursements & record reason}
}
else if (Application = Non-Doctorate Credential & Restriction 7 = True) {stop disbursements & record reason}
else If (Application = Doctorate Credential & Restriction 9 = True) {stop disbursements & record reason}

**Comment by andrewsignori-aot** [comment]
#### Life-time maximums - Full-time/Federal (#1749)
- Managed only through restrictions
- Effective date
    - Business believes that there is an effective date, but apparently, the documentation does not show it.
    - Is the goal just to avoid maximum restrictions to stop funding for in-progress disbursements?
    - Should we add “effective start” and “effective end” dates as a feature for every restriction?
- Once the restriction is in place it should stop any fed/provincial full-time disbursement (it should not impact part-time).
- Aviation, PD, non-doc, doc specific conditions must be checked for the different restrictions.
- Does it make sense to capture any sort of reason why the money was not disbursed? We do not have anything like this right now since that entire disbursement is stopped and the DB record will not even be selected for an e-Cert generation.

#### Life-time maximums - Full-time/Provincial (#1761)
- Only BC Student Loan (BCSL) is affected
- Effective date
    - Business believes that there is an effective date, but apparently, the documentation does not show it.
    - Is the goal just to avoid maximum restrictions to stop funding for in-progress disbursements?
    - Should we add “effective start” and “effective end” dates as a feature for every restriction?
- Once the restriction is in place all BC funding stops.
- Add columns to sims.disbursement_values to record the “reason”.
    - restriction_amount_subtracted
    - restriction_id_subtracted
- Save lifetime maximums per program year and also for PD, non-doc, and doc.
It can be configured using a DMN on Camunda but it must also be available on DB for estimations, for instance, during NOA the system may need to display to the Student if he is close to the maximum.
- Should we create one restriction for each PD, non-doc, and doc? If not, I do not see how it would work properly.

**Comment by andrewsignori-aot** [comment]
#### Proposed review ticket for Life-time maximums - Full-time/Federal (#1749)

- [ ] When a student gets a Full-Time Federal Lifetime Maximum restriction, it stops funding to both federal and provincial funding. Essentially, all disbursements are stopped. (this action is already set in the restrictions table as `Stop full time disbursement`).
- [ ] These restrictions are applied only to Full-time application disbursements. The related restrictions codes must have the `Stop part time disbursement` action removed: **5, 7, 9, AV**.
    - Restriction code **AV (Aviation)**: _Student has reached 17 week maximum for aviation program._
    - Restriction code **5 (PD/PPD)**: _Student has been funded for 520 weeks and has reached the lifetime maximum._
    - Restriction code **7 (non-doctorate)**: _Student has been funded for 340 weeks and has reached the lifetime maximum._
    - Restriction code **9 (doctorate)**: _Student has been funded for 400 weeks and has reached the lifetime maximum._
- [ ] Depending on the program, credential, and PD status, it is possible to have a lifetime max restriction in one scenario but still be allowed funding in a different scenario. Rules are as follows:
    - If the application is for an Aviation program, the AV restriction supersedes all other conditions.
    - If the student has PD/PDD status and the application is NOT for an Aviation program, then restriction 5, supersedes all other conditions.
    - If the student does NOT have PD/PDD status and the application is not for an Aviation program, then the applicable restriction depends on whether the application is for a doctorate or non-doctorate credentials.

According to the Restrictions Analysis file, the restriction code B5 is no longer needed when we switch to SIMS.
Logic for when and how to apply federal restriction codes
At time of ECERT generation:
if (Application = Aviation Program & Restriction AV= True) {stop disbursements & record reason}
else if (PD Status = True )
{
if (Restriction 5 = True or Restriction B7 = True) {stop disbursements & record reason}
}
else if (Application = Non-Doctorate Credential & Restriction 7 = True) {stop disbursements & record reason}
else If (Application = Doctorate Credential & Restriction 9 = True) {stop disbursements & record reason}

**Comment by andrewsignori-aot** [comment]
Proposed reviewed ticket for maximums - full-time/Provincial (#1761)

The student's cumulative full-time BCSL awards received over a lifetime need to be tracked by SIMS and identify when the student has hit the full-time BCSL Lifetime Maximum.

This will require some calculations at the time of e-Cert generation to determine if the current disbursement will cause the student to hit the maximum.  If it will, then it will be necessary to reduce the award amount accordingly so that the final total exactly equals the full-time BCSL lifetime maximum that is in effect for that program year.  

Once the maximum is hit, a restriction will be applied to the student.  All subsequent disbursements will effectively zero out the full-time BCSL award amounts at the time of ECERT generation.

**Additional Context**
 - Restriction applies only to full-time BCSL as part-time maximums follow completely different rules.
 - Only full-time BCSL is counted towards the maximum, but once the restriction is in place, all BC Funding is blocked.  This is because Grants can only be awarded to students who are eligible for Loans. 
 - The amounts used for maximums can change and need to be stored within a program year.  If a reassessment of a past application occurs, the maximum values that were in effect for the program year of the application are still in effect.  The values will never change in the middle of a program year.
- There are three different categories for maximum (PD/non-doctorate/doctorate) but the system should consider all as always the same, hence just one maximum configuration is needed and only one provincial restriction is needed.
- **NOTE:** There are times when Ministry staff allow the student to continue receiving BCSL funding beyond the lifetime maximum; they will require the ability to **IGNORE** the restriction temporarily, allowing all remaining disbursements for the current application, after which, the restriction would become **EFFECTIVE** again. (Possibly another ticket altogether)

**Logic for applying maximums**
At time of ECERT generation:
if (BCSL Lifetime Max Restriction = True) {Do not disburse BCSL and BCSG}
else if (PD/PPD status = True)
{
if ((Sum of all past full time BCSL funds received + effective BCSL award of current disbursement) => PD/PPD BCSL Max Value)
{
Reduce effective BCSL award of current disbursement so that the student hits PD/PPD BCSL Max Value exactly;
Apply BCSL Lifetime Max Restriction to a student;
Record in the database the reason for reducing the award amount.

**Acceptance Criteria**
 - [ ] New BCSL Lifetime Maximum restriction(s) are created with the appropriate actions.
 - [ ] When a disbursement causes the student to hit the lifetime maximum, the award is reduced so that the student hits the maximum value exactly.  Also, the restriction is applied to the student.
 - [ ] When there is a restriction is in place, disbursements are blocked.
 - [ ] When disbursements are blocked or reduced due to hitting the maximum, the reason is recorded.
